modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3645,12 +3672,106 @@ GDScriptParser::DataType GDScriptAnalyzer::make_global_class_meta_type(const Str
 	}
 }
 
+void GDScriptAnalyzer::update_last_expression_context(const GDScriptParser::Node *p_expression) {
+	GDScriptParser::DataType expr_type;
+	if (p_expression != nullptr) {
+		expr_type = p_expression->get_datatype();
+	}
+	last_expression_parser_ref = cache_external_class_parser(expr_type.class_type, p_expression);
+}
+
+Ref<GDScriptParserRef> GDScriptAnalyzer::cache_external_class_parser(const GDScriptParser::ClassNode *p_class, const GDScriptParser::Node *p_source) {
+	if (p_class == nullptr || parser->has_class(p_class)) {
+		return nullptr;
+	}
+
+	String script_path = p_class->get_datatype().script_path;
+	auto E = parser->depended_parsers.find(script_path);
+	if (E) {
+		// Silently try to parse.
+		E->value->raise_status(GDScriptParserRef::PARSED);
+		if (E->value->get_parser()->has_class(p_class)) {
+			// Will always be found by find_cached_parser_for_class anyway.
+			return E->value;
+		}
+	}
+
+	Ref<GDScriptParserRef> parser_ref = find_cached_parser_for_class(p_class, vformat(R"(Updating parser cache for class "%s")", p_class->fqcn), p_source);
+
+	if (parser_ref != nullptr) {
+		external_class_parser_cache.insert(p_class, parser_ref);
+	}
+
+	return parser_ref;
+}
+
+Ref<GDScriptParserRef> GDScriptAnalyzer::find_cached_parser_for_class(const GDScriptParser::ClassNode *p_class, String p_context, const GDScriptParser::Node *p_source) {
+	if (p_class == nullptr) {
+		return nullptr;
+	}
+
+	if (parser->has_class(p_class)) {
+		return nullptr;
+	}
+
+	auto E = external_class_parser_cache.find(p_class);
+	if (E) {
+		return E->value;
+	}
+
+	String script_path = p_class->get_datatype().script_path;
+
+	Ref<GDScriptParserRef> parser_ref;
+	if (parser->depended_parsers.has(script_path)) {
+		Ref<GDScriptParserRef> try_parser_ref = parser->depended_parsers[script_path];
+
+		// Silently try to parse.
+		try_parser_ref->raise_status(GDScriptParserRef::PARSED);
+		if (try_parser_ref->get_parser()->has_class(p_class)) {
+			parser_ref = try_parser_ref;
+		}
+	}
+	if (parser_ref == nullptr && last_expression_parser_ref != nullptr) {
```

Comment:
```suggestion
	if (parser_ref.is_null() && last_expression_parser_ref.is_valid()) {
```

Reply:
is comparing to nullptr not correct? it seems much more readable for a ptr-like type..

Reply:
We use the methods universally, it is the standard, it also allows for alternatives for doing the checking internally if we wish at some point

Reply:
Using these methods actually make it explicit that this is a `Ref<T>`, and not a raw pointer.

Reply:
i dont agree but ill make the change

`is_valid` is an especially confusing name, and imo `== nullptr` communicates more clearly that im checking if it is a null pointer, not a null parser ref (if i didn't know that wasn't a method)

theres no reason to know that im dealing with a `Ref`, its such a lightweight type, it only has 4 named fuctions and all have operator equivalents

its already designed to behave like a raw pointer with `->`, its at the very least, inconsistent
but i get that its like this simply because its the standard

Reply:
I mean, you'd have to argue with a lot of the codebase to deviate from the standard ;)

```
$ rg "\.is_valid()" -g'!thirdparty' | wc -l
5053
$ rg "\.is_null()" -g'!thirdparty' | wc -l
2280
```